### DISCUSSION TITLE
React에서의 상태관리란?

### DISCUSSION THREAD

**Discussion by byhhh2** [woowacourse-study/2022-woowahan-bansanghwe#120] [React]
## 1️⃣ 프론트엔드에서 상태란 무엇일까요?

- 유저와의 상호작용을 위한, 변할 수 있는 데이터
    - 유저의 눈에 보이는 데이터는 전부 상태 (화면 구성에 필요한 데이터)
    - 눈에 보이진 않지만 해당 기능을 작동시키기 위해 숨겨진 데이터도 상태
- 사용자에게 제공하고자 하는 데이터
    - 내부 로직을 위한 상태도, 간접적으로 사용자에게 노출된다.

> 한마디로,
>   
> **적절한 UI/UX를 구성하기 위한 변할 수 있는 데이터**
> 

### 📂 상태의 공유 여부

- **지역적 상태 (local state)**
    - 특정 컴포넌트 안에서만 관리되는 상태
    - ex) 다른 컴포넌트와 데이터를 공유하지 않는 form 데이터 등
- **전역적 상태 (global state)**
    - product 전체 혹은 여러 컴포넌트에서 관리되는 상태
    - 서로 다른 컴포넌트가 동일한 상태를 다룬다면 출처는 오직 한 곳(전역 공간)이어야 한다. (만약 사본이 있다면 동기화하는 과정이 더 까다롭다.)
    - ex) 장바구니에 담긴 물품의 경우 상품 선택 여부에 따라 총 주문 금액을 업데이트 해주어야 한다. (개수를 다른 컴포넌트에 전달해주어야 한다.)

### 📂 데이터 무결성

- 데이터의 정확성을 보장하기 위해 데이터 변경에 제한을 두어 안정성을 저해하는 요소를 막아, 데이터를 항상 옳게 유지하는 것
    - 무결성을 위한 방법론에는 **신뢰할 수 있는 단일 출처**가 있다. (single source of truth)

### 📂 상태가 지켜야 할 것

- **일관적인 형태를 유지해, 무결성을 지켜야 한다.**
    - 인스타로 예를 들면, 팔로워 수를 표시하는 화면이 3군데 있다고 하자, 만약 3가지 페이지에서 모든 다른 팔로워 수를 표시하는 경우 사용자는 혼란에 빠질 수 있다. 만약 팔로워 수가 변경되었다면, 3군데 모두 변경이 반영되어야 한다.
- **상태를 어디에 저장할지 적절하게 선택해야 한다.**
    - 전역 스토어, 브라우저의 스토어, 지역
    - 상태가 어디에, 어떻게 쓰이는 지에 따라 저장소를 선택해야한다.

## 2️⃣ 상태관리란 무엇일까요?

- 유저와의 상호작용을 위해 상태를 조작하고 다루는 모든 작업
- 데이터에 맞추어 적절하게 UX와 UI를 설계하고 구현하는 것
    - 예를 들면, 팔로워가 2.1억명이라고 생각해보자. 팔로워 수를 `210000000` 으로 표기하는 것은 사용자가 적절하게 이해하기도 쉽지 않고, 깨지는 UI를 제공할 수도 있다. `2.1억` 으로 표기하게 되면, 사용자의 이해도 쉬워지고 UI를 적절하게 유지할 수 있다.
- 목표한 UI/UX를 구현하기 위해 적절하게 상태를 정의, 설계하고
언제, 어떻게, 왜 변화했는지 잘 추적하는 일

### 왜 중요할까요?

- 언제든지 변할 수 있다.
    - 서비스가 방대해지면서 상태가 언제, 어디서, 왜 변화하는지 추적하기 힘들 수 있다.
    - 이런 이유로 적절한 상태관리가 필요하다.
- 부적절한 상태관리로 인해 발생하는 문제
    - 불필요한 리렌더링
    - 의도하지 않은 UI/UX
    - 유지보수 하기 힘든 코드

### 과거의 상태관리는 어땠을까?

SPA가 대중화되기 전, `data-*` 속성을 사용하여 DOM element에 직접 상태를 주입하고 사용했다. (예시 jQuery)

- 상태관리 로직이 DOM 중심으로 구성이 됐다. ⇒ 이는 상태가 언제, 어디서, 왜 변화하는지 추적하기 어렵게 한다.
- ES6 이전에는 상태를 격리하여 관리할 수 있는 선택지가 부족했고, window 객체를 사용하자니 전역오염이 걱정이 되었다. 이런 이유로 JS 보다는 DOM을 좀 더 베이스로 삼았고, DOM에 상태를 유지하는 것을 선호했다.
    - 단점
        - 상태에 대한 접근성이 떨어진다. (상태를 핸들링하기 위해서는 DOM에 직접 접근해야한다.)
        - 상태의 파편화가 심하다. (상태가 동작에 따라 여러 DOM에 흩어져 있다. 즉 디버깅할 때 살펴보아야할 로직이 많다.) 이는 상태 변화의 추적을 어렵게 한다.

> 2007년 아이폰이 출시되고, 스마트폰은 2010년 쯤 완전한 대세가 되었다. 
>  웹은 이러한 요구에 맞추어 웹앱이라고 부르는 복잡하고 정교한 웹으로 거듭나야 했다. 
>  이에 따른 SPA에 대한 요구가 생기는데 많은 상태가 생겨남으로써 상태 변화에 따른 잦은 페이지 새로고침은 나쁜 사용자 경험을 주게 된다.
> 

SPA가 대중화하고 사용자들의 요구사항도 증가하게 되었다. 

- 즉각적인 상호작용을 원하고, 더 나은 UX를 원하게 되었다. ⇒ SPA의 등장

SPA를 지원하는 여러 프레임워크는 

- DOM에 접근 없이도 데이터가 변경되면 값이 변경된다.
- 이전에 DOM 중심이던 상태관리 로직이, 데이터 중심의 상태관리 로직으로 변경된다.
- 이를 통해 어디에서 상태가 변경되었는지 추적에 유리해졌다. (접근성, 파편화 해결)

## 3️⃣ 리액트의 상태관리

- 양방향 데이터 흐름을 가지는 MVC와는 다르게 단방향 데이터 흐름을 가진다.
    - 상위 컴포넌트에서 state를 가지고, 의존된 컴포넌트가 있다면 props로 상태를 넘겨준다.
    - 만약 의존된 컴포넌트가 많다면 어떻게 될까?
        - 받은 props를 다시 props로 내려주고.. 그 props를 다시 props로 내려주게 된다.
        - 이 문제는 다시 상태관리가 어디서, 언제, 왜 변경되는지 추적하기 힘들어지게 된다. 이를 prop drilling이라고 한다.

### prop drilling을 해결하기 위해 redux 등장

- redux는 flux 아키텍처와 reducer의 개념을 합쳐서 만들어진 전역 상태관리 라이브러리

전역 상태 관리 라이브러리를 통해 

- 👍 컴포넌트들의 의존성 분리
- 👍 redux devTools를 통해 상태 변화를 추적
- 👎 많은 보일러 플레이트 코드
- 👎 API 통신 관련 코드를 전역적으로 관리

### Redux의 문제를 해결하기 위한 상태관리 라이브러리 대거 등장

- context API (라이브러리 아님)
- redux toolkit, SWR, recoil, react-query 등 ..

어떤 상태관리 라이브러리를 쓰는게 우리 서비스에 적합할까? 

### 좋은 리액트 상태관리란?

- **무분별한 전역 상태관리는 피하자.**

일반적으로 리액트에서는 지역적으로 상태관리를 하고 있는데, 이는 prop drilling을 발생시킬 수 있다. 하지만 prop drilling을 해결하기 위해 합성 컴포넌트를 활용할 수 있다. (추가로 재사용성도 좋아진다.) 

- **서버의 데이터를 전역상태에 저장할 때는 명확한 전략을 갖자.**

명확한 전략이 없다면 언제 이 데이터들을 refresh 해야하는지 파악이 힘들 수 있다. 

서버 상태를 전역상태에 저장하는 것은 보통 데이터를 캐싱한다고 하는데

어떤 시점에 이 데이터들을 동기화시켜야할지 전략이 없다면, 전역상태에 있는 데이터가 항상 최신값을 바라보지 않을 수 있다.  

이때 react-query, SWR 같은 라이브러리를 똑똑하게 활용한다면 더 좋은 API 통신 코드가 될 수 있다. 

- **상태코드는 연관 컴포넌트들과 최대한 가까이 배치하자. (state colocation)**

하나의 컴포넌트가 많은 상태를 가지고 있다면, 해당 컴포너트가 너무 많은 일을 하는 것은 아닌지 살펴보자. (우선적으로 전역상태로 빼자는 생각보다는 응집도를 높이자.)

### presentation component

- 컴포넌트는 입력이 fake data라도 표현할 수 있어야 한다.
- 어떤 데이터가 들어오는지는 상관하지 않고, 컴포넌트는 표현(presentation) 자체에 집중한다.

## 참고

- [테코톡] 온스타의 상태관리
- https://medium.com/wematch/프론트엔드의-상태관리란-무엇인가-5ff888dab7ad
- https://velog.io/@ehdgusdl9177/프론트엔드-개발에서의-상태-관리
- https://intrepidgeeks.com/tutorial/front-end-status-management-history-1

**Comment by al-bur** [comment]
react가 나오게 된 배경

**Reply by byhhh2** [reply]
왜!!!!!!!!!!!! 병민 알려조!!!

**Comment by al-bur** [comment]
javascript를 통한 DOM 조작과 react를 활용한 DOM 조작의 차이

**Comment by al-bur** [comment]
javascript를 통한 DOM 조작이 비효율적인 예시

**Reply by al-bur** [reply]
[병민 질문] javascript는 싱글 쓰레드인데, 똑같은 거 아닌가??

**Comment by al-bur** [comment]
상태란?

**Comment by al-bur** [comment]
상태와 지역 변수의 차이는?

**Comment by al-bur** [comment]
상태와 지역 변수를 사용하는 철학

**Comment by al-bur** [comment]
react query를 왜 사용하셨는지?

**Comment by al-bur** [comment]
어떻게 서버의 데이터와 동기화 된다는 건지??

**Reply by al-bur** [reply]
상태 관리 라이브러리를 사용해서, 전역에서 관리하는 거랑 똑같은 거 아닌지?

**Comment by al-bur** [comment]
캐시란?
